AX: [AX Thread Text APIs] Estimated range bounds aren't accurate

hoffmanjoshua opened the pull request:
#### 9343d8ff722ecfe6379cf2668fc1c0d13bd902ce
<pre>
AX: [AX Thread Text APIs] Estimated range bounds aren&apos;t accurate
<a href="https://bugs.webkit.org/show_bug.cgi?id=290015">https://bugs.webkit.org/show_bug.cgi?id=290015</a>
<a href="https://rdar.apple.com/147365717">rdar://147365717</a>

Reviewed by Tyler Wilcock.

Our old approach of computing text marker range bounds off of the main thread,
using a fixed, estimated character width, would lead to frames that were
inaccurate and didn&apos;t properly highlight the range. To account for variable
character widths, we can instead reference the CTFont&apos;s properties, which
we already cache on isolated objects. The `advance` of a series of the glyphs
provides an accurate horizontal width of a text range.

Follow up PRs will be necessary to address geometry and character manipulation
like transforms and letter-spacing.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::viewportRelativeFrameFromRuns):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::AXTextRuns):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::fontOrientation const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):

Canonical link: <a href="https://commits.webkit.org/292392@main">https://commits.webkit.org/292392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52534d5ec5cf4e18bf46846cd193f8945c582f60

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15466 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100911 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46360 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23898 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73098 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30325 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98855 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11803 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86587 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53426 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11534 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4334 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45696 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81703 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4437 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102940 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22919 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16715 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82142 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23170 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82606 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81501 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20445 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26082 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3532 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16256 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22882 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28037 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22541 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26020 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24282 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXTextMarker.cpp:
```diff
@@ -841,7 +841,8 @@ static FloatRect viewportRelativeFrameFromRuns(Ref<AXIsolatedObject> object, uns
     float estimatedLineHeight = relativeFrame.height() / runs->size();
-    auto runsLocalRect = runs->localRect(start, end, estimatedLineHeight);
+    CTFontRef fontRef = object->font().get();
+    auto runsLocalRect = runs->localRect(start, end, estimatedLineHeight, relativeFrame, fontRef);
```
Do we need to worry about `font()` ever being `nil`? What do we do in that case? Or can we assume we always have a font?

hoffmanjoshua replied:
I think we should add a fallback case, but we can probably just return the object's relative frame. No font would mean that the object has no `style`, which seems unlikely.

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,9 +152,18 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
+    auto computeAdvance = [&] (unsigned startIndex, unsigned endIndex) {
+        Vector<CGGlyph, 512> glyphs(endIndex - startIndex);
+        String characters = substring(startIndex, endIndex);
+        characters.convertTo16Bit();
```
Any idea why we have to do this?

hoffmanjoshua replied:
Let me see if I can remove this now that we use `characters.charactersWithNullTermination`. Originally, this string was getting converted to a 16 bit one (so that we can make `UniChar`s and we were hitting an assert there. But, `.charactersWithNullTermination` may do this implicitely.

hoffmanjoshua replied:
This is unnecessary now—removed!

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,9 +152,18 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
+        String characters = substring(startIndex, endIndex);
+        characters.convertTo16Bit();
+        auto buffer = characters.charactersWithNullTermination().value();
+        CTFontGetGlyphsForCharacters(fontRef, reinterpret_cast<const UniChar*>(buffer.data()), glyphs.data(), glyphs.size());
```
Similar code does this a bit different:

https://github.com/WebKit/WebKit/blob/f55d8061c2b360892c44ccc8355d95dbd83b93be/Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp#L62

It uses `buffer.size()` instead of `glyphs.size()`. Should we be doing that too, or as you've written it?

twilco replied:
Second question on this bit: should we be using `CTFontGetVerticalGlyphsForCharacters` for vertical text? If so, should we add a FIXME here?

hoffmanjoshua replied:
`glyphs.size()` and `buffer.size()` should be the same (i.e., the length of the string/number of characters). Maybe it's clearer to be more explitit and pass in the string's length? What do you think?

hoffmanjoshua replied:
> Second question on this bit: should we be using CTFontGetVerticalGlyphsForCharacters for vertical text? If so, should we add a FIXME here?

Yes, I'm going to investigate this a bit (see Dominic's comment above)

twilco replied:
> glyphs.size() and buffer.size() should be the same (i.e., the length of the string/number of characters). Maybe it's clearer to be more explitit and pass in the string's length? What do you think?

`buffer` is `characters`, but with `charactersWithNullTermination()` called on it. Does that mean it adds a null-terminator if necessary? If so, could that make the strings different lengths?

hoffmanjoshua replied:
Yes, this should really be `charactersWithoutNullTermination`—going to update this to use that.

hoffmanjoshua replied:
Fixed—both buffer and glyphs are guaranteed to be the same now that we use `charactersWithoutNullTermination`.

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,9 +152,18 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
     unsigned runIndexOfLargerOffset = indexForOffset(largerOffset);
 
+    auto computeAdvance = [&] (unsigned startIndex, unsigned endIndex) {
+        Vector<CGGlyph, 512> glyphs(endIndex - startIndex);
```
Should we add a `RELEASE_ASSERT(endIndex >= startIndex)`? Or do you feel super confident in the calling code of this lambda?

hoffmanjoshua replied:
I feel confident since we only call this in three spots—but a release assert couldn't hurt.

minorninth commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -152,9 +152,18 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
+        auto buffer = characters.charactersWithNullTermination().value();
+        CTFontGetGlyphsForCharacters(fontRef, reinterpret_cast<const UniChar*>(buffer.data()), glyphs.data(), glyphs.size());
+        Vector<CGSize> advances(glyphs.size(), CGSizeZero);
+        return CTFontGetAdvancesForGlyphs(fontRef, kCTFontOrientationDefault, glyphs.begin(), advances.begin(), glyphs.size());
```
I think kCTFontOrientationDefault probably means horizontal, this will probably need to be adjusted for vertical text

hoffmanjoshua replied:
Hmm right—we'll either need to read this from somewhere, but more likely, have to cache whether the font is vertical or horizontal here.

hoffmanjoshua replied:
Updated this to take in a dynamic orientation property that we now cache. I've added a fixme, since the current run-index loop only considers building text bounds horizontally, so I'll need to follow up to fix this for vertical font orientations as well.

twilco commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -173,19 +182,14 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
+                if (smallerOffset - offsetOfFirstCharacterInRun > 0)
+                    widthPriorToStart = computeAdvance(offsetOfFirstCharacterInRun, smallerOffset);
+
+                unsigned endOffsetInLine = runIndexOfSmallerOffset == runIndexOfLargerOffset ? largerOffset : !i ? run.text.length() : runLengthSumTo(i - 1) + run.text.length();
```
Would it be too clunky to split this up into multiple lines rather than using a double-ternary? Might be easier to parse overall that way 🤔

twilco commented on Source/WebCore/accessibility/AXTextRun.h:
```diff
@@ -100,19 +101,11 @@ struct AXTextRuns {
-    // APIs like AXBoundsForTextMarkerRangeAttribute quickly off the main-thread, and get more accurate
-    // sizes on-demand for runs that assistive technologies are actually requesting these APIs from.
-    static constexpr uint8_t defaultEstimatedCharacterWidth = 12;
-    uint8_t estimatedCharacterWidth { defaultEstimatedCharacterWidth };
```
Glad to remove this!

minorninth commented on Source/WebCore/accessibility/AXTextRun.cpp:
```diff
@@ -173,19 +182,14 @@ FloatRect AXTextRuns::localRect(unsigned start, unsigned end, float lineHeight)
-                measuredWidth = fullRunWidth - widthPriorToStart - widthAfterEnd;
+                unsigned widthPriorToStart = 0;
+                if (smallerOffset - offsetOfFirstCharacterInRun > 0)
+                    widthPriorToStart = computeAdvance(offsetOfFirstCharacterInRun, smallerOffset);
```
Can you convince me that we're not doing O(n^2) work here in the worst case?

My concern is that in a run with n characters, we've got a for loop of length n, but for each iteration of the loop we're calling computeAdvance which needs to iterate over all of the characters in its range to return the correct answer.

Is this a concern or in practice are we avoiding recomputing the same character multiple times?

hoffmanjoshua replied:
The loop that this is inside iterates through the text runs, not through the characters in the run, so we won't iterate over each character twice. 

```c++
for (unsigned i = 0; i <= runIndexOfLargerOffset; i++) {
```

Each time we enter a new loop, we are only looking at the characters in that particular line/run, so work isn't duplicated there.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
```diff
@@ -630,6 +630,7 @@ void AXIsolatedObject::setProperty(AXProperty propertyName, AXPropertyValueVaria
 #if ENABLE(AX_THREAD_TEXT_APIS)
         [](AXTextRuns& runs) { return !runs.size(); },
         [](RetainPtr<CTFontRef>& typedValue) { return !typedValue; },
+        [](CTFontOrientation typedValue) { return typedValue == kCTFontOrientationDefault; },
```
Here's an alternate idea that would save a bit of memory. Why don't we make `kCTFontOrientationHorizontal` the default, and only store a value in the map if it is `kCTFontOrientationVertical` or `kCTFontOrientationDefault`? Horizontal text is probably the most common orientation. What do you think?

hoffmanjoshua replied:
Yeah we can definitely do this, we just need to add some new attribute getters since `Horizontal` isn't the compilers's default value for this enum.

minorninth replied:
Maybe it'd be easier to just store BOOL isVertical as the property rather than the actual CTFontOrientation?

hoffmanjoshua replied:
This is outdated now—I've updated this to store the internal type, `FontOrientation`, which is just a bool wrapper.

hoffmanjoshua committed:
AX: [AX Thread Text APIs] Estimated range bounds aren't accurate
https://bugs.webkit.org/show_bug.cgi?id=290015
rdar://147365717

Reviewed by Tyler Wilcock.

Our old approach of computing text marker range bounds off of the main thread,
using a fixed, estimated character width, would lead to frames that were
inaccurate and didn't properly highlight the range. To account for variable
character widths, we can instead reference the CTFont's properties, which
we already cache on isolated objects. The `advance` of a series of the glyphs
provides an accurate horizontal width of a text range.

Follow up PRs will be necessary to address geometry and character manipulation
like transforms and letter-spacing.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AXTextMarker.cpp:
(WebCore::viewportRelativeFrameFromRuns):
* Source/WebCore/accessibility/AXTextRun.cpp:
(WebCore::AXTextRuns::localRect const):
* Source/WebCore/accessibility/AXTextRun.h:
(WebCore::AXTextRuns::AXTextRuns):
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AccessibilityObjectCocoa.mm:
(WebCore::AccessibilityObject::fontOrientation const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):

Canonical link: https://commits.webkit.org/292392@main
